pkg/app/piped/executor/analysis/metrics_analyzer.go

```diff
@@ -114,6 +142,149 @@ func (a *metricsAnalyzer) analyzeWithThreshold(ctx context.Context) (expected bo
 		return false, nil
 	}
 
-	a.logPersister.Successf("[%s] The query result is expected one. Performed query: %q", a.id, a.cfg.Query)
 	return true, nil
 }
+
+// Return false if primary deviates in the specified direction compared to the previous deployment.
+// Return an error if the evaluation could not be executed normally.
+// elapsedTime is used to compare metrics at the same point in time after the analysis has started.
+func (a *metricsAnalyzer) analyzeWithPrevious(ctx context.Context) (bool, error) {
+	now := time.Now()
+	queryRange := metrics.QueryRange{
+		From: now.Add(-a.cfg.Interval.Duration()),
+		To:   now,
+	}
+	points, err := a.provider.QueryPoints(ctx, a.cfg.Query, queryRange)
+	if err != nil {
+		return false, fmt.Errorf("failed to run query: %w", err)
+	}
+
+	prevMetadata, err := a.analysisResultStore.GetLatestAnalysisResult(ctx)
+	if err != nil {
+		return false, fmt.Errorf("failed to fetch the most recent successful analysis metadata: %w", err)
+	}
+	// Compare it with the previous metrics when the same amount of time as now has passed since the start of the stage.
+	elapsedTime := time.Since(a.stageStartTime)
+	prevNow := time.Unix(prevMetadata.StartTime, 0).Add(elapsedTime)
```

Comment:
nit: `preStart`

Reply:
Not sure why do we need to add the `elapsedTime`?
Is `preMetadata.StartTime` is the time when the ANALYSIS stage at previous deployment started, right?

Reply:
This `analyzeWithPrevious` is called periodically. To be fair, as commented in the code, we need to compare two metrics that have had the same amount of time elapsed.

Reply:
>nit: preStart

If anything, `preEnd` is preferred. But technically incorrect. `preTo` is the most exact name but a bit difficult to understand...

Reply:
Let's say the duration is `15m` and interval is `5m` Piped analyzes three times. Each time, we need to compare like:
| | 1st evaluation | 2nd evaluation | 3rd evaluation |
| ---- | ---- | ---- | ---- |
| prev| 0m ~ 5m | 6m ~10m | 11m ~ 15m |
| now | 0m ~5m | 6m ~ 10m | 11m ~ 15m |

Ask me anything if any ;)

Reply:
> This analyzeWithPrevious is called periodically
Ah, got it.

Reply:
`prevFrom` and `prevTo` look good.

Reply:
That looks cool